spec/cspec/AARCH64/Kernel_C.thy

```diff
@@ -0,0 +1,110 @@
+(*
+ * Copyright 2023, Proofcraft Pty Ltd
+ *
+ * SPDX-License-Identifier: GPL-2.0-only
+ *)
+
+theory Kernel_C
+imports
+  "ExecSpec.MachineTypes"
+  "CLib.CTranslationNICTA"
+  "AsmRefine.CommonOps"
+begin
+
+external_file
+  "../c/build/$L4V_ARCH/kernel_all.c_pp"
+
+context begin interpretation Arch .
+
+requalify_types
+  machine_state
+
+end
+
+declare [[populate_globals=true]]
+
+context begin interpretation Arch . (*FIXME: arch_split*)
+
+type_synonym cghost_state = "(machine_word \<rightharpoonup> vmpage_size) * (machine_word \<rightharpoonup> nat)
+    * ghost_assertions"
+
+definition
+  gs_clear_region :: "addr \<Rightarrow> nat \<Rightarrow> cghost_state \<Rightarrow> cghost_state" where
+  "gs_clear_region ptr bits gs \<equiv>
+   (%x. if x \<in> {ptr..+2 ^ bits} then None else fst gs x,
+    %x. if x \<in> {ptr..+2 ^ bits} then None else fst (snd gs) x, snd (snd gs))"
+
+definition
+  gs_new_frames:: "vmpage_size \<Rightarrow> addr \<Rightarrow> nat \<Rightarrow> cghost_state \<Rightarrow> cghost_state"
+  where
+  "gs_new_frames sz ptr bits \<equiv> \<lambda>gs.
+   if bits < pageBitsForSize sz then gs
+   else (\<lambda>x. if \<exists>n\<le>mask (bits - pageBitsForSize sz).
+                  x = ptr + n * 2 ^ pageBitsForSize sz then Some sz
+             else fst gs x, snd gs)"
+
+definition
+  gs_new_cnodes:: "nat \<Rightarrow> addr \<Rightarrow> nat \<Rightarrow> cghost_state \<Rightarrow> cghost_state"
+  where
+  "gs_new_cnodes sz ptr bits \<equiv> \<lambda>gs.
+   if bits < sz + 4 then gs
+   else (fst gs, \<lambda>x. if \<exists>n\<le>mask (bits - sz - 4). x = ptr + n * 2 ^ (sz + 4)
+                     then Some sz
+                     else fst (snd gs) x, snd (snd gs))"
+
+abbreviation
+  gs_get_assn :: "int \<Rightarrow> cghost_state \<Rightarrow> machine_word"
+  where
+  "gs_get_assn k \<equiv> ghost_assertion_data_get k (snd o snd)"
+
+abbreviation
+  gs_set_assn :: "int \<Rightarrow> machine_word \<Rightarrow> cghost_state \<Rightarrow> cghost_state"
+  where
+  "gs_set_assn k v \<equiv> ghost_assertion_data_set k v (apsnd o apsnd)"
+
+declare [[record_codegen = false]]
+declare [[allow_underscore_idents = true]]
+
+end
+
+(* workaround for the fact that the C parser wants to know the vmpage sizes*)
+(* create appropriately qualified aliases *)
+context begin interpretation Arch . global_naming vmpage_size
+requalify_consts ARMSmallPage ARMLargePage ARMHugePage
+end
+
+definition
+  ctcb_size_bits :: nat
+where
+  "ctcb_size_bits \<equiv> 11"
```

Comment:
I checked against the latest Kernel_C on aarch64 crefine branch and it's 10, because I got it wrong in this version. Consider pulling in that one? 04ea4e5468c1f3d9d20b41b31fb816335a3f4c0a

Reply:
(probably I squashed that change in at some point, but I definitely pushed it out)

Reply:
Will do. We should really make that symbolic at some point, but today is not that day.